src/psyclone/tests/utils.py

```diff
@@ -28,3 +50,141 @@ def count_lines(root, string_name):
+
+
+def walk(parent, my_type):
+    ''' recurse through a tree of objects and return those that are
+    instances of mytype. Assumes that the children of any given node
+    are held in the list node.children '''
+    local_list = []
+    for child in parent.children:
+        if isinstance(child, my_type):
+            local_list.append(child)
+        local_list += walk(child, my_type)
+    return local_list
+
+
+def find_fortran_file(path, root_name):
+    ''' Returns the full path to a Fortran source file. Searches for
+    files with suffixes defined in FORTRAN_SUFFIXES. Raises IOError
+    if no matching file is found. '''
+    name = os.path.join(path, root_name)
+    for suffix in FORTRAN_SUFFIXES:
+        if os.path.isfile(str(name)+"."+suffix):
+            name += "." + suffix
+            return name
+    raise IOError("Cannot find a Fortran file '{0}' with suffix in {1}".
+                  format(name, FORTRAN_SUFFIXES))
+
+
+def compile_file(filename, f90, f90flags):
+    ''' Compiles the specified Fortran file into an object file (in
+    the current working directory) using the specified Fortran compiler
+    and flags. Raises a CompileError if the compilation fails. '''
+
+    # Build the command to execute
+    if f90flags:
+        arg_list = [f90, f90flags, '-c', filename]
+    else:
+        arg_list = [f90, '-c', filename]
+
+    # Attempt to execute it using subprocess
+    import subprocess
+    try:
+        build = subprocess.Popen(arg_list,
+                                 stdout=subprocess.PIPE,
+                                 stderr=subprocess.STDOUT)
+        (output, error) = build.communicate()
+    except OSError as err:
+        print "Failed to run: {0}: ".format(" ".join(arg_list))
+        print "Error was: ", str(err)
+        raise CompileError(str(err))
+
+    # Check the return code
+    stat = build.returncode
+    if stat != 0:
+        print output
+        if error:
+            print "========="
+            print error
+        raise CompileError(output)
+    else:
+        return True
+
+
+def code_compiles(api, psy_ast, tmpdir, f90, f90flags):
+    '''Attempts to build the Fortran code supplied as an AST of
+    f2pygen objects. Returns True for success, False otherwise.
+    If no Fortran compiler is available then returns True. All files
+    produced are deleted. '''
+    from psyclone import f2pygen
+
+    # API-specific set-up - where to find infrastructure source files
+    # and which ones to build
+    supported_apis = ["dynamo0.3"]
+    if api not in supported_apis:
+        raise CompileError("Unsupported API in code_compiles. Got {0} but "
+                           "only support {1}".format(api, supported_apis))
+
+    if api == "dynamo0.3":
+        base_path = os.path.join(os.path.dirname(os.path.abspath(__file__)),
+                                 "test_files", "dynamo0p3")
+        from dynamo0p3_build import INFRASTRUCTURE_MODULES as module_files
+
+    kernel_modules = []
```

Comment:
I think kernel_modules can be found simply by psy_gen.invokes.names. This would remove the need for the psy_module change to dynamo0p3.py and psyGen.py and the walk function above

Reply:
Well spotted. It's not quite as simple as you suggest - I have to dig down to the kernel objects and then get their associated `module_name` - but it's nicer.